cpp/src/arrow/array/validate.cc

```diff
@@ -38,195 +39,172 @@ namespace internal {
+      return Status::Invalid("Fixed size list has negative list size");
     }
+
     int64_t expected_values_length = -1;
-    if (MultiplyWithOverflow(len, value_size, &expected_values_length) ||
-        array.values()->length() != expected_values_length) {
-      return Status::Invalid("Values Length (", array.values()->length(),
-                             ") is not equal to the length (", len,
-                             ") multiplied by the value size (", value_size, ")");
+    if (MultiplyWithOverflow(data.length, list_size, &expected_values_length) ||
+        values.length != expected_values_length) {
+      return Status::Invalid("Values length (", values.length,
+                             ") is not equal to the length (", data.length,
+                             ") multiplied by the value size (", list_size, ")");
+    }
+
+    const Status child_valid = ValidateArray(values);
+    if (!child_valid.ok()) {
+      return Status::Invalid("Fixed size list child array invalid: ",
+                             child_valid.ToString());
     }
 
     return Status::OK();
   }
 
-  Status Visit(const StructArray& array) {
-    const auto& struct_type = checked_cast<const StructType&>(*array.type());
-    // Validate fields
-    for (int i = 0; i < array.num_fields(); ++i) {
-      // array.field() may crash due to an assertion in ArrayData::Slice(),
-      // so check invariants before
-      const auto& field_data = *array.data()->child_data[i];
-      if (field_data.length < array.offset()) {
-        return Status::Invalid("Struct child array #", i,
-                               " has length smaller than struct array offset (",
-                               field_data.length, " < ", array.offset(), ")");
-      }
+  Status Visit(const StructType& type) {
+    for (int i = 0; i < type.num_fields(); ++i) {
+      const auto& field_data = *data.child_data[i];
 
-      auto it = array.field(i);
-      if (it->length() != array.length()) {
+      // Validate child first, to catch nonsensical length / offset etc.
+      const Status field_valid = ValidateArray(field_data);
+      if (!field_valid.ok()) {
         return Status::Invalid("Struct child array #", i,
-                               " has length different from struct array (", it->length(),
-                               " != ", array.length(), ")");
+                               " invalid: ", field_valid.ToString());
       }
 
-      auto it_type = struct_type.field(i)->type();
-      if (!it->type()->Equals(it_type)) {
+      if (field_data.length < data.length + data.offset) {
         return Status::Invalid("Struct child array #", i,
-                               " does not match type field: ", it->type()->ToString(),
-                               " vs ", it_type->ToString());
+                               " has length smaller than struct array (",
+                               field_data.length, " != ", data.length + data.offset, ")");
```

Comment:
Users might get confused that what we're implying is the struct array's length includes its offset:
```suggestion
                               " has length smaller than the slice viewed by struct array (",
                               field_data.length, " < ", data.length + data.offset, ")");
```

Reply:
I'll commit an alternate suggestion.